bd-crash-handler/src/uploader.rs

```diff
@@ -0,0 +1,282 @@
+#[cfg(test)]
+#[path = "./uploader_test.rs"]
+mod tests;
+
+use async_trait::async_trait;
+use bd_api::upload::{IntentDecision, TrackedArtifactIntent, TrackedArtifactUpload};
+use bd_api::DataUpload;
+use bd_client_common::file::{read_compressed_protobuf, write_compressed_protobuf};
+use bd_client_common::file_system::FileSystem;
+use bd_proto::protos::client::api::{UploadArtifactIntentRequest, UploadArtifactRequest};
+use bd_proto::protos::client::artifact::artifact_upload_index::Artifact;
+use bd_proto::protos::client::artifact::ArtifactUploadIndex;
+use bd_shutdown::ComponentShutdown;
+use bd_time::{OffsetDateTimeExt, TimeProvider};
+use mockall::automock;
+use std::collections::VecDeque;
+use std::path::PathBuf;
+use std::sync::{Arc, LazyLock};
+use uuid::Uuid;
+
+/// Root directory for all files used for storage and uploading.
+pub static REPORT_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| "report_uploads".into());
+
+/// The index file used for tracking all of the individual files.
+pub static REPORT_INDEX_FILE: LazyLock<PathBuf> = LazyLock::new(|| "report_index.pb".into());
+
+type NewUpload = (Uuid, Vec<u8>);
+
+#[automock]
+#[async_trait]
+pub trait Client: Send + Sync {
+  async fn enqueue_upload(&self, contents: Vec<u8>) -> anyhow::Result<Uuid>;
+}
+
+pub struct UploadClient {
+  upload_tx: tokio::sync::mpsc::Sender<NewUpload>,
+}
+
+#[async_trait]
+impl Client for UploadClient {
+  /// Dispatches a payload to be uploaded, returning the associated artifact UUID.
+  async fn enqueue_upload(&self, contents: Vec<u8>) -> anyhow::Result<Uuid> {
```

Comment:
nit: async not needed if we are doing try_send which makes sense to avoid overflow/blocking.